Sources/xcproj/PBXProjObjects+Helpers.swift

```diff
@@ -15,7 +16,182 @@ public extension PBXProj.Objects {
+        return groups.first(where: {
+            children.contains($0.key) && ($0.value.name == groupName || $0.value.path == groupName)
+        })
+    }
+
+    /// Adds new group with the give name to the given parent group.
+    /// Group name can be a path with components separated by `/`.
+    /// Will create new groups for intermediate path components or use existing groups.
+    /// Returns all new or existing groups in the path and their references.
+    ///
+    /// - Parameters:
+    ///   - groupName: group name, can be a path with components separated by `/`
+    ///   - toGroup: parent group
+    ///   - options: additional options, default is empty set.
+    /// - Returns: all new or existing groups in the path and their references.
+    public func addGroup(named groupName: String, to toGroup: PBXGroup, options: GroupAddingOptions = []) -> [(reference: String, group: PBXGroup)] {
+        return addGroups(groupName.components(separatedBy: "/"), to: toGroup, options: options)
+    }
+
+    private func addGroups(_ groupNames: [String], to toGroup: PBXGroup, options: GroupAddingOptions) -> [(reference: String, group: PBXGroup)] {
+        guard !groupNames.isEmpty else { return [] }
+        let newGroup = createOrGetGroup(named: groupNames[0], in: toGroup, options: options)
+        return [newGroup] + addGroups(Array(groupNames.dropFirst()), to: newGroup.1, options: options)
+    }
+
+    private func createOrGetGroup(named groupName: String, in parentGroup: PBXGroup, options: GroupAddingOptions) -> (reference: String, group: PBXGroup) {
+        if let existingGroup = group(named: groupName, inGroup: parentGroup) {
+            return existingGroup
+        }
+
+        let newGroup = PBXGroup(
+            children: [],
+            sourceTree: .group,
+            name: groupName,
+            path: options.contains(.withoutFolder) ? nil : groupName
+        )
+        let reference = generateReference(newGroup, groupName)
+        addObject(newGroup, reference: reference)
+        parentGroup.children.append(reference)
+        return (reference, newGroup)
+    }
+
+    /// Adds file at the give path to the project or returns existing file and its reference.
+    ///
+    /// - Parameters:
+    ///   - filePath: path to the file.
+    ///   - sourceRoot: path to project's source root.
+    /// - Returns: new or existing file and its reference.
+    public func addFile(at filePath: Path, sourceRoot: Path) throws -> (reference: String, file: PBXFileReference) {
+        guard filePath.isFile else { throw XCodeProjEditingError.notAFile(path: filePath) }
+
+        if let existingFileReference = fileReferences.first(where: {
+            filePath == fullPath(fileElement: $0.value, reference: $0.key, sourceRoot: sourceRoot)
+        }) {
+            return (existingFileReference.key, existingFileReference.value)
+        }
+
+        let fileReference = PBXFileReference(
+            sourceTree: .absolute,
```

Comment:
Adding absolute path is very restrictive. This will restrict the usage of this project to single path in a filesystem. It would be useful only for situations where project is always generated.

I recommend rethinking this helper function.
My suggestion:
`addFile` function will take filePath and optional PBXGroup where nil will be a fallback to mainGroup.

Reply:
@allu22 indeed! I improved this method by adding group and source tree parameters. Source tree is required to calculate initial value for file path property, as it will be different for different source tree values, i.e. it can be relative to group or to project. Also added helper method to calculate relative paths. I made it public as this method can be used for example to update file path when moving it to a different group.
Some of the source tree values are not supported in file path calculation (file will be added with nil path), but we can extend it later, it feels that use cases for these values should be rare.